### PR TITLE
Fix loss of mask sigma clipping (issue #17401)

### LINKED ISSUE #17401: Bug introduced by PR #17086 deleting mask for MaskedArray input
### Description

As a result of PR #17086, the mask of a `MaskedArray` input is lost in circumstances where the changed code is executed, producing incorrect output results. This is because the array is copied with `np.copy(data)` instead of `data.copy()`, which will use the `np.ma.copy()` function if `data` is a `MaskedArray`. The fix is straightforward and I'm working on a PR but I need to set everything up and write a test, etc.

### Expected behavior

In this example, errant data have been masked so the statistics should only be calculated for the elements with value 1, and hence the mean should be 1.0 and the stddev 0.0. (The `grow` parameter is included here since, as stated in PR #17086, this part of the code isn't executed otherwise.)
```
>>> import numpy as np
>>> from astropy.stats import sigma_clipped_stats
>>> x = np.array([1,1,1,1,1,1,1,1,5,5]).astype(float)
>>> y = np.ma.masked_where(x > 1, x)
>>> print(sigma_clipped_stats(y, grow=1))
(1.0, 1.0, 0.0)
```

### How to Reproduce

The actual results show that the two values of 5 are erroneously included in the stats.
```
>>> import numpy as np
>>> from astropy.stats import sigma_clipped_stats
>>> x = np.array([1,1,1,1,1,1,1,1,5,5]).astype(float)
>>> y = np.ma.masked_where(x > 1, x)
>>> print(sigma_clipped_stats(y, grow=1))
(1.8, 1.0, 1.6)
```

### Versions

```python
import astropy
try:
    astropy.system_info()
except AttributeError:
    import platform; print(platform.platform())
    import sys; print("Python", sys.version)
    import astropy; print("astropy", astropy.__version__)
    import numpy; print("Numpy", numpy.__version__)
    import erfa; print("pyerfa", erfa.__version__)
    try:
        import scipy
        print("Scipy", scipy.__version__)
    except ImportError:
        print("Scipy not installed")
    try:
        import matplotlib
        print("Matplotlib", matplotlib.__version__)
    except ImportError:
        print("Matplotlib not installed")
```
```
macOS-14.1.1-x86_64-i386-64bit
Python 3.10.12 | packaged by conda-forge | (main, Jun 23 2023, 22:39:40) [Clang 15.0.7 ]
astropy 6.1.6
Numpy 1.25.2
pyerfa 2.0.1.4
Scipy 1.13.0
Matplotlib 3.7.2
```

### PR DESCRIPTION
**Pull request by chris-simpson**
### Description

This pull request is to address ...

Fixes #17401 by using the appropriate `copy()` method of the input data, whether it's an array or masked array.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Review by mhvk — APPROVED** [review]
Really quite bad that by default `np.copy()` doesn't properly copy a masked array (looks like it would need `subok=True`), but this a clean way to avoid it.

Approving, though not merging, as @larrybradley should have a final look.

File: astropy/stats/tests/test_sigma_clipping.py [hunk 1]
```diff
@@ -696,3 +696,11 @@ def nan_mad_std(data, axis=None):
     )
 
     assert_allclose(result1, result2)
+
+
+def test_propagation_of_mask():
```
Comment (mhvk) [inline astropy/stats/tests/test_sigma_clipping.py:701]: Actually, there is a test failure here, because `grow` requires scipy. You need to add `@pytest.mark.skipif(not HAS_SCIPY, reason="test requires scipy")` (assuming the test strictly needs it; if it also does the test without `grow`, I think that would be preferred).

Comment (jehturner) [reply astropy/stats/tests/test_sigma_clipping.py:701]: Thanks @mhvk! I believe the correct solution is to add the test annotation, since @chris-simpson  noted in https://github.com/astropy/astropy/issues/17401 that the grow parameter is needed to reproduce the problem. I am unable to push to Chris's branch, however, in order to set the tests running (or even send him a PR for some reason), so I'll have to wait for him to add that...

Comment (jehturner) [reply astropy/stats/tests/test_sigma_clipping.py:701]: BTW, I think this is quite a significant regression, because (eg.) it will change the behaviour of `FittingWithOutlierRemoval` when passed `sigma_clip` as the `outlier_func`, which is typical usage.

**Comment by jehturner** [comment]
@chris-simpson, I haven't seen the failing check before, but my best guess is that it wants a new line at the end of 17402.bugfix.rst.

File: astropy/stats/tests/test_sigma_clipping.py (hunk 1 above)
Comment (larrybradley) [reply astropy/stats/tests/test_sigma_clipping.py:702]: You can reproduce the issue with `sigma_clipped_stats(y, cenfunc=np.mean, axis=0)` without using the `grow` keyword.  You need a non-default cenfunc or stdfunc and an axis that is not `None`.

**Comment by pllim** [comment]
pre-commit.ci autofix

File: astropy/stats/tests/test_sigma_clipping.py (hunk 1 above)
Comment (pllim) [reply astropy/stats/tests/test_sigma_clipping.py:702]: Re: regression -- Does that mean we should not backport?

Comment (jehturner) [reply astropy/stats/tests/test_sigma_clipping.py:702]: No, I just meant that the bug addressed by this PR seems fairly significant, although I see that you already have this down for the next patch releases (and presumably it won't make 7.0, which is unfortunate, but we'll just have to constrain DRAGONS as incompatible with 7.0.0 etc.). Thanks!

Comment (pllim) [reply astropy/stats/tests/test_sigma_clipping.py:702]: I think a `!=7.0.0` would automatically pull in the next bug fix.

tl;dr -- Does this also mean you don't have this problem in 6.x series?

Comment (pllim) [reply astropy/stats/tests/test_sigma_clipping.py:702]: Ah nvm, looking at milestone of offending PR linked, I think you also need a `!=6.1.5`.

Comment (jehturner) [reply astropy/stats/tests/test_sigma_clipping.py:702]: Yes, that, and also `!=6.1.6`. Having the milestones defined is very helpful in that regard (otherwise we'd have to put `<=6.1.4`).

Comment (pllim) [reply astropy/stats/tests/test_sigma_clipping.py:702]: For the 6.x pins, not sure if it matters because now that v7 is out, I think technically we're not obligated to release another 6.x anymore but maybe @astropy/astropy-project-release-team can confirm.

Comment (pllim) [reply astropy/stats/tests/test_sigma_clipping.py:702]: (But for consistency, we should still backport to v6.1.x branch, just in case.)

Comment (saimn) [reply astropy/stats/tests/test_sigma_clipping.py:702]: 7.0 is not yet out so still a bit of time to backport a fix.

Comment (larrybradley) [reply astropy/stats/tests/test_sigma_clipping.py:702]: This test could be rewritten to avoid requiring `scipy` to be installed, but better is the enemy of good here --  I'd rather get this in the 7.0 release.  I'll go ahead and merge.

**Review by larrybradley — APPROVED** [review]
Thanks, @chris-simpson

**Comment by jehturner** [comment]
Thank you! Did you want an additional test for `cenfunc=np.mean, axis=0`? I believe the same code with those args would do it (probably a cut & paste job, as it doesn't seem worth factoring out the 2 lines in common?).

@saimn, @pllim: would you mind confirming that this definitely won't go into 7.0? Just because it's now in that branch and 7.0 hasn't been tagged yet, so it's not really obvious without being familiar with the mechanics of how finalize it.

**Comment by pllim** [comment]
Actually I think it will land in v7.0.0 since it is already backported to v7.0.x but the release isn't cut yet.

**Comment by pllim** [comment]
A way to find out is to test against the dev wheel tomorrow (this patch should be in by then).

**Comment by saimn** [comment]
v7.0.0 is out with this fix. Should we do a v6.1.7 release as well (with just this fix) or it is fine for you @chris-simpson @jehturner to exclude v6.1.5/6 ? 
cc @astrofrog

**Comment by saimn** [comment]
v6.1.7 is released.

**Comment by jehturner** [comment]
Oh, thank you, @saimn, that's great!
